### PR TITLE
More behavioral compatibility changes

### PR DESCRIPTION
**Pull request by kgaillot**

### CODE REVIEW

**Comment by kgaillot** [comment]
@nrwahl2 , here's another compatibility PR when you get a chance. specific tests good

File: lib/common/tests/rules/pcmk__evaluate_attr_expression_test.c [hunk 1]
```diff
@@ -660,6 +661,18 @@ op_eq_passes(void **state)
...
 }
 
+#define EXPR_EMPTY_EQ_EMPTY_PASSES                      \
+        "<" PCMK_XE_EXPRESSION " " PCMK_XA_ID "='e' "   \
+        PCMK_XA_ATTRIBUTE "='empty' "                   \
+        PCMK_XA_OPERATION "='" PCMK_VALUE_EQ "' "       \
+        PCMK_XA_VALUE "='' />"
+
+static void
+op_empty_eq_empty_passes(void **state)
```
Comment (nrwahl2) [inline lib/common/tests/rules/pcmk__evaluate_attr_expression_test.c:671]: we've been using `op_<op>_<description>` for the test naming, doesn't matter much

File: lib/common/tests/rules/pcmk__evaluate_attr_expression_test.c [hunk 2]
```diff
@@ -673,6 +686,19 @@ op_eq_fails(void **state)
...
 }
 
+#define EXPR_UNDEFINED_EQ_EMPTY_FAILS                   \
+        "<" PCMK_XE_EXPRESSION " " PCMK_XA_ID "='e' "   \
+        PCMK_XA_ATTRIBUTE "='boo' "                     \
+        PCMK_XA_OPERATION "='" PCMK_VALUE_EQ "' "       \
+        PCMK_XA_VALUE "='' />"
+
+static void
+op_undefined_eq_empty_fails(void **state)
```
Comment (nrwahl2) [inline lib/common/tests/rules/pcmk__evaluate_attr_expression_test.c:696]: ditto, a bit more counterintuitive here in that the name looks like the op is `not_defined` (we have other `op_undefined` tests)

File: lib/pacemaker/pcmk_sched_location.c [hunk 3]
```diff
@@ -165,16 +165,11 @@ generate_location_rule(pcmk_resource_t *rsc, xmlNode *rule_xml,
...
 
-        default:
-            /* @COMPAT When we can break behavioral backward compatibility,
-             * return false
-             */
-            pcmk__config_warn("Location constraint rule %s has invalid "
-                              PCMK_XA_BOOLEAN_OP " value '%s', using default "
-                              "'" PCMK_VALUE_AND "'",
+        default: // Not possible with schema validation enabled
+            pcmk__config_warn("Ignoring location constraint rule %s "
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_sched_location.c:169]: Should this become `pcmk__config_err()`?

File: lib/common/watchdog.c [hunk 4]
```diff
@@ -94,24 +94,32 @@ panic_local(void)
...
 
-    if (pcmk__starts_with(panic_action, "sync-")) {
+    if (pcmk__starts_with(full_panic_action, "sync-")) {
+        panic_action += sizeof("sync-") - 1;
         sync();
-        panic_action += strlen("sync-");
-    };
+    }
+
+    if (pcmk__str_empty(panic_action)
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/watchdog.c:96]: good point

File: lib/common/watchdog.c [hunk 5]
```diff
@@ -94,24 +94,32 @@ panic_local(void)
 
     /* We're either pacemakerd, or a pacemaker daemon running as root */
 
-    if (pcmk__starts_with(panic_action, "sync-")) {
+    if (pcmk__starts_with(full_panic_action, "sync-")) {
+        panic_action += sizeof("sync-") - 1;
         sync();
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/watchdog.c:93]: it's fine, the user clearly indicated they wanted a sync

File: lib/common/watchdog.c [hunk 6]
```diff
@@ -42,6 +42,34 @@ sysrq_trigger(char t)
...
+{
+    if (ppid > 1) { // pacemakerd is still our parent
+        crm_emerg("Escalating panic to " PCMK__SERVER_PACEMAKERD "[%lld]",
+                  (long long) ppid);
+    } else { // Signal (non-parent) pacemakerd if possible
+#if HAVE_LINUX_PROCFS
+        union sigval signal_value;
+
+        ppid = pcmk__procfs_pid_of(PCMK__SERVER_PACEMAKERD);
+        crm_emerg("Signaling " PCMK__SERVER_PACEMAKERD "[%lld] to panic",
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/watchdog.c:64]: Worth handling

For bundles, pacemaker-remoted is PID 1. pacemakerd would be PID 1 in a container if someone containerized Pacemaker :)

File: lib/common/watchdog.c [hunk 7]
```diff
@@ -79,22 +79,15 @@ panic_local_nonroot(pid_t ppid)
...
     const char *full_panic_action = pcmk__env_option(PCMK__ENV_PANIC_ACTION);
     const char *panic_action = full_panic_action;
+    int reboot_cmd = RB_AUTOBOOT; // Default panic action is reboot
 
-    // Default panic action is to reboot
-    int reboot_cmd = RB_AUTOBOOT;
-
-    if (uid != 0) { // Non-root caller such as the controller
-        panic_local_nonroot(ppid);
+    if (geteuid() > 0) { // Non-root caller such as the controller
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/watchdog.c:86]: good catch

File: lib/pacemaker/pcmk_sched_location.c (hunk 3 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_location.c:169]: yes

**Comment by kgaillot** [comment]
updated commits per review

**Review by nrwahl2 — APPROVED** [review]

**Pull request merged**